genienlp/ned/base.py

```diff
@@ -0,0 +1,333 @@
+        while i < len(tokens):
+            token = tokens[i]
+            # sort by number of tokens so longer keys get matched first
+            matched_items = sorted(self.all_aliases.keys(token), key=lambda item: len(item), reverse=True)
+            found = False
+            for key in matched_items:
+                type = self.alias2type[key]
+                key_tokenized = key.split()
+                cur = i
+                j = 0
+                while cur < len(tokens) and j < len(key_tokenized):
+                    if tokens[cur] != key_tokenized[j]:
+                        break
+                    j += 1
+                    cur += 1
+
+                if j == len(key_tokenized):
+                    if is_banned(' '.join(key_tokenized)):
+                        continue
+
+                    # match found
+                    found = True
+                    tokens_type_ids.extend([[self.typeqid2id[type] * self.max_features_size] for _ in range(i, cur)])
+
+                    # move i to current unprocessed position
+                    i = cur
+                    break
+
+            if not found:
+                tokens_type_ids.append([self.unk_id * self.max_features_size])
+                i += 1
+
+        return tokens_type_ids
+
+    def lookup_longer(self, tokens):
+        i = 0
+        tokens_type_ids = []
+
+        length = len(tokens)
+        found = False
+        while i < length:
+            end = length
+            while end > i:
+                tokens_str = ' '.join(tokens[i:end])
+                if tokens_str in self.all_aliases:
+                    # match found
+                    found = True
+                    tokens_type_ids.extend(
+                        [[self.typeqid2id[self.alias2type[tokens_str]] * self.max_features_size] for _ in range(i, end)]
+                    )
+                    # move i to current unprocessed position
+                    i = end
+                    break
+                else:
+                    end -= 1
+            if not found:
+                tokens_type_ids.append([self.unk_id * self.max_features_size])
+                i += 1
+            found = False
+
+        return tokens_type_ids
+
+    def lookup_entities(self, tokens, entities):
+        tokens_type_ids = [[self.unk_id] * self.max_features_size] * len(tokens)
+        tokens_text = " ".join(tokens)
+
+        for ent in entities:
+            if ent not in self.all_aliases:
+                continue
+            ent_num_tokens = len(ent.split(' '))
+            idx = tokens_text.index(ent)
+            token_pos = len(tokens_text[:idx].strip().split(' '))
+            type = self.typeqid2id.get(self.alias2type[ent], self.unk_id)
+            tokens_type_ids[token_pos : token_pos + ent_num_tokens] = [[type] * self.max_features_size] * ent_num_tokens
+
+        return tokens_type_ids
+
+    def lookup(self, tokens, database_lookup_method=None, min_entity_len=2, max_entity_len=4):
```

Comment:
This method is not used by Bootleg right? So it doesn't belong here. It should go into a subclass.

Reply:
Ok yeah I should create a base class for Naive and Oracle and many of these methods should move there.